cumulusci/tasks/salesforce/update_profile.py

```diff
@@ -47,168 +65,182 @@ def _init_options(self, kwargs):
+            self.package_xml_path = os.path.join(
+                CUMULUSCI_PATH, "cumulusci", "files", "admin_profile.xml"
             )
-        )
-        api_retrieve = ApiRetrieveUnpackaged(
-            self,
-            self._get_retrieve_package_xml_content(),
-            self.project_config.project__package__api_version,
-        )
-        unpackaged = api_retrieve()
-        unpackaged.extractall(self.retrieve_dir)
-
-    def _process_metadata(self):
-        self.logger.info(f"Processing retrieved metadata in {self.retrieve_dir}")
-        path = self.retrieve_dir / "profiles" / f"{self.profile_name}.profile"
-        self.tree = elementtree_parse_file(path)
-
-        self._set_apps_visible()
-        self._set_classes_enabled()
-        self._set_fields_editable()
-        self._set_fields_readable()
-        self._set_pages_enabled()
-        self._set_tabs_visibility()
-        self._set_record_types()
-
-        self.tree.write(
-            path, "utf-8", xml_declaration=True, default_namespace=self.namespaces["sf"]
+
+    def freeze(self, step):
+        # Preserve behavior from when we subclassed Deploy.
+
+        steps = super().freeze(step)
+        for step in steps:
+            if step["kind"] == "other":
+                step["kind"] = "metadata"
+        return steps
+
+    def _generate_package_xml(self, operation):
+        if operation is MetadataOperation.RETRIEVE:
+            with open(self.package_xml_path, "r") as f:
+                package_xml_content = f.read()
+
+            package_xml_content = package_xml_content.format(**self.namespace_prefixes)
+
+            if (
+                self.options["include_packaged_objects"]
+                or "package_xml" not in self.options
+            ):
+                # We need to rewrite the package.xml for one or two reasons.
+                # Either we are using packaged-object expansion, or we're using
+                # the built-in admin_profile.xml and need to substitute in
+                # profile API names.
+                package_xml = metadata_tree.fromstring(package_xml_content)
+
+                if self.options["include_packaged_objects"]:
+                    self._expand_package_xml(package_xml)
+                if "package_xml" not in self.options:
+                    self._expand_profile_members(package_xml)
+
+                package_xml_content = package_xml.tostring(xml_declaration=True)
+
+            return package_xml_content
+        else:
+            return super()._generate_package_xml(operation)
+
+    def _expand_profile_members(self, package_xml):
+        profile_names = package_xml.find("types", name="Profile")
+        for profile in self.api_names:
+            profile_names.append("members", text=profile)
+
+    def _expand_package_xml(self, package_xml):
+        # Query the target org for all namespaced objects
+        # Add these entities to the package.xml
+
+        results = self.tooling.query(
```

Comment:
I think we need to use query_all in case there is more than one page of results

Reply:
Yeah, you're right. I wish `simple_salesforce` did not use `query_all` to mean something other than "include deleted records" (which is what it means at the API level).

Reply:
yep. I wonder if we can goad @prescod into writing a better Salesforce REST API client as his next weekend project. :-p

Reply:
We should look at https://github.com/forcedotcom/SalesforcePy

Not least because we could probably get contributor rights.